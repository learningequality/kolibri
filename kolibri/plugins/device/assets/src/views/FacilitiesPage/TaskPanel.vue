<template>

  <div class="task-panel" :class="{ 'task-panel-sm': windowIsSmall }">
    <div class="icon">
      <transition mode="out-in">
        <KCircularLoader
          v-if="showCircularLoader || taskIsCanceling"
          :size="24"
          :stroke="5"
        />
        <KIcon
          v-else-if="taskIsFailed"
          icon="helpNeeded"
          :style="{ fill: $themeTokens.error }"
        />
        <KIcon
          v-else-if="taskIsCompleted"
          icon="done"
          :style="{ fill: $themeTokens.success }"
        />
        <KIcon
          v-else
          icon="inProgress"
          :style="{ fill: $themeTokens.annotation }"
        />
      </transition>
    </div>

    <div class="details">
      <p
        class="default-margin details-status"
        :style="{ color: $themeTokens.annotation }"
      >
        {{ statusMsg }}
      </p>

      <h2 v-if="headingMsg" class="default-margin details-description">
        {{ headingMsg }}
      </h2>

      <p v-if="underHeadingMsg" class="default-margin fs0">
        {{ underHeadingMsg }}
      </p>

      <div v-if="loaderType && !statusHidesLoader" class="details-progress-bar">
        <template v-if="loaderType === 'determinate'">
          <KLinearLoader
            class="k-linear-loader"
            type="determinate"
            :delay="false"
            :progress="task.percentage * 100"
            :style="{ backgroundColor: $themeTokens.fineLine }"
          />
          <span v-if="taskPercentage" class="details-percentage">
            {{ $formatNumber(taskPercentage, { style: 'percent' }) }}
          </span>
        </template>

        <template v-else-if="loaderType === 'indeterminate'">
          <KLinearLoader
            class="k-linear-loader"
            type="indeterminate"
            :delay="false"
            :style="{ backgroundColor: $themeTokens.fineLine }"
          />
        </template>
      </div>

      <p v-if="underProgressMsg" class="default-margin fs0">
        {{ underProgressMsg }}
      </p>

      <p v-if="startedByMsg" class="default-margin fs0" :style="{ color: $themeTokens.annotation }">
        {{ startedByMsg }}
      </p>
    </div> 

    <div class="buttons" :class="{ 'button-lift': Boolean(loaderType) }">
      <KButton
        v-if="buttonSet === 'cancel'"
        :disabled="taskIsCanceling"
        :text="coreString('cancelAction')"
        appearance="flat-button"
        @click="$emit('cancel')"
      />
      <KButton
        v-if="buttonSet === 'clear' || buttonSet === 'retry'"
        :text="coreString('clearAction')"
        appearance="flat-button"
        @click="$emit('clear')"
      />
      <KButton
        v-if="buttonSet === 'retry'"
        :text="coreString('retryAction')"
        appearance="raised-button"
        primary
        @click="$emit('retry')"
      />
    </div>
  </div>

</template>


<script>

  import responsiveWindowMixin from 'kolibri.coreVue.mixins.responsiveWindowMixin';
  import commonCoreStrings from 'kolibri.coreVue.mixins.commonCoreStrings';
  import commonTaskStrings from 'kolibri.coreVue.mixins.commonTaskStrings';
  import { TaskStatuses } from '../../constants';

  export default {
    name: 'TaskPanel',
    mixins: [commonCoreStrings, responsiveWindowMixin, commonTaskStrings],
    props: {
      task: {
        type: Object,
        required: true,
      },
      statusMsg: {
        type: String,
        required: true,
      },
      headingMsg: {
        type: String,
        required: true,
      },
      underHeadingMsg: {
        type: String,
        required: false,
      },
      underProgressMsg: {
        type: String,
        required: false,
      },
      showCircularLoader: {
        type: Boolean,
        default: false,
      },
      loaderType: {
        type: String,
        required: false,
        validator(value) {
          return value === 'determinate' || value === 'indeterminate';
        },
      },
      buttonSet: {
        type: String,
        required: false,
        validator(value) {
          return value === '' || value === 'clear' || value === 'cancel' || value === 'retry';
        },
      },
    },
    computed: {
      taskIsCompleted() {
        return this.task.status === TaskStatuses.COMPLETED;
      },
      taskIsCanceling() {
        return this.task.status === TaskStatuses.CANCELING;
      },
      taskIsCanceled() {
        return this.task.status === TaskStatuses.CANCELED;
      },
      taskIsFailed() {
        return this.task.status === TaskStatuses.FAILED || this.taskIsCanceled;
      },
      taskPercentage() {
        return this.task.percentage;
      },
      startedByMsg() {
        if (!this.task.started_by_username) {
          return '';
        }
        return this.getTaskString('taskStartedByLabel', {
          username: this.task.started_by_username || this.getTaskString('unknownUsername'),
        });
      },
      statusHidesLoader() {
        return (
          this.task.status === TaskStatuses.PENDING ||
          this.taskIsCompleted ||
          this.taskIsCanceling ||
          this.taskIsFailed
        );
      },
    },
  };

</script>


<style lang="scss" scoped>

  $fs0: 12px;
  $fs1: 14px;
  $fs2: 16px;

  .default-margin {
    margin: 8px 0;
  }

  .icon {
    padding: 0 16px;

    .task-panel-sm & {
      align-self: flex-start;
    }
  }

  .icon svg {
    width: 24px;
    height: 24px;
  }

  .task-panel {
    display: flex;
    align-items: center;
  }

  .task-panel-sm {
    flex-direction: column;
    padding-top: 16px;
    padding-bottom: 16px;
  }

  .details {
    flex-grow: 1;
    padding: 16px;

    .task-panel-sm & {
      padding-top: 0;
      padding-bottom: 0;
    }
  }

  .details-description {
    font-size: $fs2;
  }

  .details-progress-bar {
    display: flex;
    align-items: center;
    max-width: 450px;
    margin-bottom: 16px;
  }

  // CSS overrides for linear loader
  .k-linear-loader {
    height: 10px;

    /deep/ .ui-progress-linear-progress-bar {
      height: 100%;
    }
  }

  .details-percentage {
    // min-width ensures num % stay on same line
    min-width: 48px;
    margin-left: 16px;
    font-size: $fs1;
  }

  .details-status {
    font-size: $fs0;
  }

  .details-size {
    font-size: $fs1;
  }

  .fs0 {
    font-size: $fs0;
  }

  .buttons {
    .task-panel-sm & {
      align-self: flex-end;
    }
  }

  .buttons-lift {
    // Lift button a little to align with progress bar
    margin-top: -24px;
  }

</style>
