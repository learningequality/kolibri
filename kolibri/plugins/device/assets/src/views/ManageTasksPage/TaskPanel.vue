<template>

  <div class="task-panel" :class="{ 'task-panel-sm': windowIsSmall }">
    <div class="icon">
      <transition mode="out-in">
        <KIcon
          v-if="taskIsFailed"
          icon="helpNeeded"
          :style="{ fill: $themeTokens.error }"
        />
        <KIcon
          v-else-if="taskIsCompleted"
          icon="done"
          :style="{ fill: $themeTokens.success }"
        />
        <KCircularLoader
          v-else-if="taskIsRunning"
          :size="24"
          :stroke="5"
        />
        <KIcon
          v-else
          icon="inProgress"
          :style="{ fill: $themeTokens.annotation }"
        />
      </transition>
    </div>

    <div class="details">
      <p class="details-status" :style="{ color: $themeTokens.annotation }">
        {{ statusText }}
      </p>
      <h2 class="details-description">
        {{ descriptionText }}
      </h2>
      <div v-if="taskIsRunning || taskIsCanceling" class="details-progress-bar">
        <template v-if="taskIsRunning">
          <KLinearLoader
            class="k-linear-loader"
            :type="loaderType"
            :delay="false"
            :progress="task.percentage * 100"
            :style="{ backgroundColor: $themeTokens.fineLine }"
          />
          <span v-if="taskPercentage" class="details-percentage">
            {{ $formatNumber(taskPercentage, { style: 'percent' }) }}
          </span>
        </template>
        <template v-else-if="taskIsCanceling">
          <KLinearLoader
            class="k-linear-loader"
            type="indeterminate"
            :delay="false"
            :style="{ backgroundColor: $themeTokens.fineLine }"
          />
        </template>
      </div>
      <template v-if="!taskIsCompleted">
        <p v-if="sizeText" class="details-size">
          {{ sizeText }}
        </p>
      </template>
      <template v-else>
        <p v-if="finishedSizeText" class="details-size">
          {{ finishedSizeText }}
        </p>
      </template>
      <p class="details-startedby" :style="{ color: $themeTokens.annotation }">
        {{ startedByText }}
      </p>
    </div>

    <div class="buttons" :class="{ 'button-lift': taskIsRunning }">
      <KButton
        v-if="taskIsCancellable || taskIsClearable"
        :disabled="taskIsCanceling"
        :text="buttonLabel"
        appearance="flat-button"
        @click="handleClick"
      />
    </div>
  </div>

</template>


<script>

  // Displays a single Task and its metadata, and provides buttons
  // to cancel or clear it.

  import responsiveWindowMixin from 'kolibri.coreVue.mixins.responsiveWindowMixin';
  import commonCoreStrings from 'kolibri.coreVue.mixins.commonCoreStrings';
  import bytesForHumans from 'kolibri.utils.bytesForHumans';

  import { TaskStatuses, TaskTypes } from '../../constants';

  const taskToTrMap = {
    [TaskTypes.REMOTECONTENTIMPORT]: 'importChannelPartial',
    [TaskTypes.DISKCONTENTIMPORT]: 'importChannelPartial',
    [TaskTypes.REMOTEIMPORT]: 'importChannelWhole',
    [TaskTypes.DISKIMPORT]: 'importChannelWhole',
    [TaskTypes.DISKEXPORT]: 'exportChannelWhole',
    [TaskTypes.DISKCONTENTEXPORT]: 'exportChannelPartial',
    [TaskTypes.DELETECHANNEL]: 'deleteChannelWhole',
    [TaskTypes.UPDATECHANNEL]: 'updatingChannelVersion',
  };

  const taskToTrPrefixMap = {
    [TaskTypes.REMOTECONTENTIMPORT]: 'import',
    [TaskTypes.DISKCONTENTIMPORT]: 'import',
    [TaskTypes.UPDATECHANNEL]: 'import',
    [TaskTypes.REMOTEIMPORT]: 'import',
    [TaskTypes.DISKIMPORT]: 'import',
    [TaskTypes.DISKEXPORT]: 'export',
    [TaskTypes.DISKCONTENTEXPORT]: 'export',
    [TaskTypes.DELETECHANNEL]: 'delete',
  };

  const statusToTrMap = {
    [TaskStatuses.COMPLETED]: 'statusComplete',
    [TaskStatuses.FAILED]: 'statusFailed',
    [TaskStatuses.RUNNING]: 'statusInProgress',
    [TaskStatuses.QUEUED]: 'statusInQueue',
    [TaskStatuses.CANCELED]: 'statusCanceled',
    [TaskStatuses.CANCELING]: 'statusCanceling',
  };

  export default {
    name: 'TaskPanel',
    mixins: [commonCoreStrings, responsiveWindowMixin],
    props: {
      task: {
        type: Object,
        required: true,
      },
    },
    computed: {
      buttonLabel() {
        if (this.taskIsClearable) {
          return this.coreString('clearAction');
        }
        return this.coreString('cancelAction');
      },
      taskIsRunning() {
        return this.task.status === TaskStatuses.RUNNING;
      },
      taskIsCompleted() {
        return this.task.status === TaskStatuses.COMPLETED;
      },
      taskIsCanceling() {
        return this.task.status === TaskStatuses.CANCELING;
      },
      taskIsCanceled() {
        return this.task.status === TaskStatuses.CANCELED;
      },
      taskIsFailed() {
        return this.task.status === TaskStatuses.FAILED || this.taskIsCanceled;
      },
      taskIsCancellable() {
        return this.task.cancellable;
      },
      taskIsClearable() {
        return this.task.clearable;
      },
      taskPercentage() {
        if (this.task.database_ready === false) {
          return null;
        }
        return this.task.percentage;
      },
      loaderType() {
        // Show an indeterminate loader while bulk import is still in pre-importcontent step.
        if (this.taskPercentage === null) {
          return 'indeterminate';
        }
        return 'determinate';
      },
      descriptionText() {
        let trName = taskToTrMap[this.task.task];
        if (
          this.task.task === TaskTypes.DELETECHANNEL &&
          ((this.task.node_ids && this.task.node_ids.length) ||
            (this.task.exclude_node_ids && this.task.exclude_node_ids.length))
        ) {
          trName = 'deleteChannelPartial';
        }
        return this.$tr(trName, {
          channelName: this.task.channel_name || this.$tr('unknownChannelName'),
          newVersion: this.task.new_version,
        });
      },
      sizeText() {
        const { file_size, total_resources } = this.task;
        if (file_size && total_resources) {
          return this.$tr('numResourcesAndSize', {
            numResources: total_resources,
            bytesText: bytesForHumans(file_size),
          });
        } else if (file_size) {
          return this.$tr('cancelSize', {
            bytesText: bytesForHumans(file_size),
          });
        }
        return '';
      },
      finishedSizeText() {
        const {
          transferred_file_size,
          transferred_resources,
          file_size,
          total_resources,
        } = this.task;
        // Special case for canceled exports
        if (
          (this.task.task === TaskTypes.DISKEXPORT ||
            this.task.task === TaskTypes.DISKCONTENTEXPORT) &&
          this.task.status === TaskStatuses.CANCELED
        ) {
          return '';
        }
        if (file_size && total_resources) {
          const trPrefix = taskToTrPrefixMap[this.task.task];
          if (
            transferred_file_size &&
            transferred_resources &&
            (transferred_file_size < file_size || transferred_resources < total_resources)
          ) {
            return this.$tr(`${trPrefix}PartialRatio`, {
              currentResources: transferred_resources,
              totalResources: total_resources,
              currentSize: bytesForHumans(transferred_file_size),
              totalSize: bytesForHumans(file_size),
            });
          } else {
            return this.$tr(`${trPrefix}Success`, {
              totalResources: total_resources,
              totalSize: bytesForHumans(file_size),
            });
          }
        }
        return '';
      },
      statusText() {
        const trName = statusToTrMap[this.task.status];
        return this.$tr(trName);
      },
      startedByText() {
        return this.$tr('startedByUser', {
          user: this.task.started_by_username || this.$tr('unknownUsername'),
        });
      },
    },
    methods: {
      handleClick() {
        if (this.taskIsCompleted || this.taskIsFailed) {
          this.$emit('clickclear');
        } else {
          this.$emit('clickcancel');
        }
      },
    },
    $trs: {
      startedByUser: {
        message: "Started by '{user}'",
        context: 'Refers to the content management *task*.',
      },
      numResourcesAndSize: {
        message:
          '{numResources} {numResources, plural, one {resource} other {resources}} ({bytesText})',
        context: 'Indicates the number of resources and their size.',
      },
      cancelSize: {
        message: 'Exported size: ({bytesText})',
        context: 'Indicates the number of resources and their size.',
      },

      statusInProgress: {
        message: 'In-progress',
        context: 'Label indicating that a task is in progress.',
      },
      statusInQueue: {
        message: 'Waiting',
        context: 'Label indicating that a task is queued.\n',
      },
      statusComplete: {
        message: 'Finished',
        context: 'Label indicating that the *task* was completed successfully.',
      },
      statusFailed: {
        message: 'Failed',
        context: 'Label indicating that a task failed, i.e. it has not been completed.',
      },
      statusCanceled: {
        message: 'Canceled',
        context: 'Refers to a canceled task in the task manager section.',
      },
      statusCanceling: {
        message: 'Canceling',
        context: 'Refers to a task being canceled in the task manager section.',
      },
      importChannelWhole: {
        message: `Import '{channelName}'`,
        context:
          "Indicates the user is importing an entire channel. For example:\n\n'Import 'Ubongo Kids''",
      },
      importChannelPartial: {
        message: `Import resources from '{channelName}'`,
        context:
          'Indicates the user is importing a selection of resources from a specific channel.\n',
      },
      exportChannelWhole: {
        message: `Export '{channelName}'`,
        context: 'Indicates the user is exporting an entire channel.',
      },
      exportChannelPartial: {
        message: `Export resources from '{channelName}'`,
        context:
          'Indicates the user is exporting a selection of resources from a specific channel.',
      },
      deleteChannelWhole: {
        message: `Delete '{channelName}'`,
        context: 'Refers to deleting an entire channel.',
      },
      /* eslint-disable kolibri/vue-no-unused-translations */
      deleteChannelPartial: {
        message: `Delete resources from '{channelName}'`,
        context: 'Indicates the channel from which resources will be deleted.',
      },
      /* eslint-enable */
      updatingChannelVersion: {
        message: `Update {channelName} to version {newVersion}`,
        context: 'Refers to a task where a channel is updated to a new version.',
      },
      // Catch-all strings if the channel or username doesn't get attached to Task
      unknownUsername: {
        message: 'Unknown user',
        context: 'Displays if the name of the user carrying out a specific task is not known.',
      },
      unknownChannelName: {
        message: '(Channel name unavailable)',
        context: 'This displays if Kolibri is unable to recognize the name of the channel.',
      },
      /* eslint-disable kolibri/vue-no-unused-translations */
      // stubs
      stopAction: {
        message: 'Stop',
        context: 'Button to stop a task from executing.',
      },
      importPartialRatio: {
        message:
          '{currentResources} of {totalResources} {totalResources, plural, one {resource} other {resources}} ({currentSize} of {totalSize}) imported',
        context:
          "Indicates how many resources are being imported out of the total resources available on that specific channel and the size of the resources. For example:\n\n'718 of 742 resources (22 GB of 22 GB) imported'",
      },
      exportPartialRatio: {
        message:
          '{currentResources} of {totalResources} {totalResources, plural, one {resource} other {resources}} ({currentSize} of {totalSize}) exported',
        context:
          "Indicates how many resources are being exported out of the total resources available on that specific channel, and the size of the resources. For example:\n\n'7 of 10 resources (22 GB of 22 GB) exported'",
      },
      deletePartialRatio: {
        message:
          '{currentResources} of {totalResources} {totalResources, plural, one {resource} other {resources}} ({currentSize} of {totalSize}) deleted',
        context:
          "Indicates how many resources are being deleted out of the total resources available on that specific channel. For example:\n\n'10 of 20 resources deleted'",
      },
      importSuccess: {
        message:
          '{totalResources} {totalResources, plural, one {resource} other {resources}} ({totalSize}) successfully imported',
        context:
          "Indicates the successful import of a specified number of resources. For example:\n\n'4752 resources (29 GB) successfully imported'",
      },
      exportSuccess: {
        message:
          '{totalResources} {totalResources, plural, one {resource} other {resources}} ({totalSize}) successfully exported',
        context:
          "Indicates the successful export of a specified number of resources.\n\n'7 of 10 resources successfully exported'",
      },
      deleteSuccess: {
        message:
          '{totalResources} {totalResources, plural, one {resource} other {resources}} ({totalSize}) successfully deleted',
        context:
          "Indicates the successful deletion of a specified number of resources. For example:\n\n'976 resources (178 MB) successfully deleted'",
      },
      /* eslint-enable kolibri/vue-no-unused-translations */
    },
  };

</script>


<style lang="scss" scoped>

  $fs0: 12px;
  $fs1: 14px;
  $fs2: 16px;

  p,
  h2 {
    margin: 8px 0;
  }

  .icon {
    padding: 0 16px;

    .task-panel-sm & {
      align-self: flex-start;
    }
  }

  .icon svg {
    width: 24px;
    height: 24px;
  }

  .task-panel {
    display: flex;
    align-items: center;
  }

  .task-panel-sm {
    flex-direction: column;
    padding-top: 16px;
    padding-bottom: 16px;
  }

  .details {
    flex-grow: 1;
    width: 100%;
    padding: 16px;

    .task-panel-sm & {
      padding-top: 0;
      padding-bottom: 0;
    }
  }

  .details-description {
    font-size: $fs2;
  }

  .details-progress-bar {
    display: flex;
    align-items: center;
    max-width: 450px;
    margin-bottom: 16px;
  }

  // CSS overrides for linear loader
  .k-linear-loader {
    height: 10px;

    /deep/ .ui-progress-linear-progress-bar {
      height: 100%;
    }
  }

  .details-percentage {
    // min-width ensures num % stay on same line
    min-width: 48px;
    margin-left: 16px;
    font-size: $fs1;
  }

  .details-status {
    font-size: $fs0;
  }

  .details-size {
    font-size: $fs1;
  }

  .details-startedby {
    font-size: $fs0;
  }

  .buttons {
    .task-panel-sm & {
      align-self: flex-end;
    }
  }

  .buttons-lift {
    // Lift button a little to align with progress bar
    margin-top: -24px;
  }

</style>
