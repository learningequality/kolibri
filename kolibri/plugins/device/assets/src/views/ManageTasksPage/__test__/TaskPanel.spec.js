import { mount } from '@vue/test-utils';
import TaskPanel from '../TaskPanel';

function makeWrapper(propsData) {
  const wrapper = mount(TaskPanel, { propsData });
  return { wrapper };
}

describe('TaskPanel', () => {
  const exportTask = {
    type: 'DISKCONTENTEXPORT',
    status: 'CANCELED',
    clearable: true,
    channel_name: 'Canceled disk export channel test',
    started_by_username: 'Tester',
    file_size: 5000,
    total_resources: 500,
  };

  it('renders correctly when it is a canceled DISKCONTENTEXPORT task', () => {
    const { wrapper } = makeWrapper({ task: exportTask });
    // File size/resource numbers should not be shown for canceled exports
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders correctly when it is a canceled DISKEXPORT task (bulk export)', () => {
    const { wrapper } = makeWrapper({ task: { ...exportTask, type: 'DISKEXPORT' } });
    expect(wrapper.html()).toMatchSnapshot();
  });
});
