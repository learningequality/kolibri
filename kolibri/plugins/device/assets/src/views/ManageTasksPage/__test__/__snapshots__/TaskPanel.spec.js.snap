// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskPanel renders correctly when it is a canceled DISKCONTENTEXPORT task 1`] = `
<div class="task-panel">
  <div class="icon">
    <mat-svg name="error" category="alert" style="fill: #d32f2f;" mode="out-in"></mat-svg>
  </div>
  <div class="details">
    <p class="details-status" style="color: rgb(97, 97, 97);">
      Canceled
    </p>
    <h2 class="details-description">
      Export resources from 'Canceled disk export channel test'
    </h2>
    <!---->
    <!---->
    <p class="details-startedby" style="color: rgb(97, 97, 97);">
      Started by 'Tester'
    </p>
  </div>
  <div class="buttons"><button dir="auto" type="button" tabindex="0" class="button KButton-noKey-0_37nh0f"><span>Clear</span>
      <!----></button></div>
</div>
`;

exports[`TaskPanel renders correctly when it is a canceled DISKEXPORT task (bulk export) 1`] = `
<div class="task-panel">
  <div class="icon">
    <mat-svg name="error" category="alert" style="fill: #d32f2f;" mode="out-in"></mat-svg>
  </div>
  <div class="details">
    <p class="details-status" style="color: rgb(97, 97, 97);">
      Canceled
    </p>
    <h2 class="details-description">
      Export 'Canceled disk export channel test'
    </h2>
    <!---->
    <!---->
    <p class="details-startedby" style="color: rgb(97, 97, 97);">
      Started by 'Tester'
    </p>
  </div>
  <div class="buttons"><button dir="auto" type="button" tabindex="0" class="button KButton-noKey-0_37nh0f"><span>Clear</span>
      <!----></button></div>
</div>
`;
